File: javascript/Frequency Counter/isAnagram.js
//Frequency Counter Method to find Anagram in linear time O(n)
// Alternate solution O(nlogn) sort the strings

function isAnagram(str1, str2) {
  if (str1.length != str2.length) return false;

  lookup = {};

  for (let i in str1) {
    let letter = str1[i];
    lookup[letter] ? (lookup[letter] += 1) : (lookup[letter] = 1);
  }

  for (let i in str2) {
    let letter = str2[i];

    if (!lookup[letter]) {
      return false;
    } else {
      lookup[letter] -= 1;
    }
  }

  return true;
}

str1 = "eat";
str2 = "ate";

console.log(isAnagram(str1, str2));

File: python/Patterns/Frequency-Counter/isAnagram.py
def isAnagram(str1,str2):

    lookup = {}
    for i in str1:
        if i in lookup:
            lookup[i]+=1
        else:
            lookup[i] =1
    
    for i in str2:
        if i not in lookup:
            False
        else:
            lookup[i]-=1

    return True

str1 = "eat"
str2 = "ate"
print(isAnagram(str1, str2))